### ISSUE TITLE
CCM is in CrashLoopBackOff on Flatcar flavor

### ISSUE THREAD

**Issue by invidian** [kubernetes-sigs/cluster-api-provider-azure#3387]
/kind bug

**What steps did you take and what happened:**

Create a cluster using Flatcar flavor. CCM logs produce the following errors:
```sh
W0404 09:47:50.730535       1 azure_zones.go:51] syncRegionZonesMap: error when get zones: Retriable: true, RetryAfter: 0s, HTTPStatusCode: -1, RawError: azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/d38033ba-ec21-470c-96cf-4c6db9658d8b/providers?api-version=2020-06-01: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Post "https://login.microsoftonline.com/f41c056a-c993-42d0-8d91-57f0ff222694/oauth2/token": tls: failed to verify certificate: x509: certificate signed by unknown authority', will retry after 30m0s
E0404 09:47:50.730549       1 azure.go:651] InitializeCloudFromConfig: failed to sync regional zones map for the first time: Retriable: true, RetryAfter: 0s, HTTPStatusCode: -1, RawError: azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/d38033ba-ec21-470c-96cf-4c6db9658d8b/providers?api-version=2020-06-01: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Post "https://login.microsoftonline.com/f41c056a-c993-42d0-8d91-57f0ff222694/oauth2/token": tls: failed to verify certificate: x509: certificate signed by unknown authority'
F0404 09:47:50.730721       1 controllermanager.go:315] Cloud provider azure could not be initialized: could not init cloud provider azure: Retriable: true, RetryAfter: 0s, HTTPStatusCode: -1, RawError: azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/d38033ba-ec21-470c-96cf-4c6db9658d8b/providers?api-version=2020-06-01: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Post "https://login.microsoftonline.com/f41c056a-c993-42d0-8d91-57f0ff222694/oauth2/token": tls: failed to verify certificate: x509: certificate signed by unknown authority'
```

**What did you expect to happen:**

All controlplane workloads to be stable.

**Anything else you would like to add:**

PR adding tests for this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3386

**Environment:**

- cluster-api-provider-azure version: 93443faf985733167903668179772c54fd230ada
- Kubernetes version: (use `kubectl version`): v1.26.3
- OS (e.g. from `/etc/os-release`): Flatcar Container Linux by Kinvolk 3374.2.5 (Oklo)

**Comment by invidian** [comment]
Yeah, so CCM mounts `/etc/ssl` from the host for CA bundle, but on Flatcar `/etc/ssl/openssl.cnf` is a symlink to `../../usr/share/ssl/openssl.cnf`, which of course is not there which makes it fail. I'll check with Flatcar team what's the recommended way forward in this case.

**Comment by invidian** [comment]
I wonder if this is a regression in https://github.com/kubernetes-sigs/cluster-api-provider-azure/commit/3e4da2823246799bb77ee2be0d1e12bc4f6b1cfc#diff-2c7fc9c9c12632c5e550abeb31614356abd17710d948a8664f9d2f2163e38b0dL277 and Azure CCM was never tested on Flatcar.

**Comment by invidian** [comment]
Yeah, it seems the problem is only how CCM is now installed. This patch fixes it:
```patch
$ git diff test/
diff --git test/e2e/cloud-provider-azure.go test/e2e/cloud-provider-azure.go
index a7f207f8..8f50444d 100644
--- test/e2e/cloud-provider-azure.go
+++ test/e2e/cloud-provider-azure.go
@@ -58,6 +58,10 @@ func InstallCalicoAndCloudProviderAzureHelmChart(ctx context.Context, input clus
                options.StringValues = append(options.StringValues, fmt.Sprintf("cloudNodeManager.imageTag=%s", os.Getenv("IMAGE_TAG_CNM")))
        }

+       if input.ConfigCluster.Flavor == "flatcar" {
+               options.StringValues = append(options.StringValues, "cloudControllerManager.caCertDir=/usr/share/ca-certificates")
+       }
+
        clusterProxy := input.ClusterProxy.GetWorkloadCluster(ctx, input.ConfigCluster.Namespace, input.ConfigCluster.ClusterName)
        InstallHelmChart(ctx, clusterProxy, defaultNamespace, cloudProviderAzureHelmRepoURL, cloudProviderAzureChartName, cloudProviderAzureHelmReleaseName, options)

```

@nawazkh can you try adding this to your PR to see if it fixes the e2e tests?

**Comment by nawazkh** [comment]
> can you try adding this to your PR to see if it fixes the e2e tests?

I added it to my PR and running tests (ref: f8ad961ef46fe856eda6d96056751e8b7e4a068f)
Does it also mean that the flatcar template (under flavors) needs to be updated if the tests pass?

**Comment by invidian** [comment]
>Does it also mean that the flatcar template (under flavors) needs to be updated if the tests pass?

I'm not sure. I don't see any reference of CCM in the flavor templates. Docs says to manually install CCM on workload cluster as far as I see, so it seems to me kind of out of scope. We could maybe document that Flatcar requires different `caCertDir`, but I'm not sure it belongs to CAPZ.

**Comment by bavarianbidi** [comment]
@invidian we had a similar issue with the ca-mount when using flatcar (but on `openstack`):

tl;dr: we've dropped the `/etc/ssl` mount from the host into the CCM

https://github.com/kubernetes/cloud-provider-openstack/pull/1928

**Comment by CecileRobertMichon** [comment]
@invidian @nawazkh what's the status on this issue?

**Comment by invidian** [comment]
The fix is included in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3386, but I was not able to figure out why e2e tests are still failing on CI, as mentioned in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3386#issuecomment-1515136573 :confused:

**Comment by nawazkh** [comment]
Would you be open to pair sometime to take a closer look at flatcar test? I am not sure why the flatcar tests are failing too. I will also follow up via slack.
